eventprocessor/eventprocessor.go

```diff
@@ -0,0 +1,107 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/honeycombio/honeycomb-lambda-extension/extension"
+	"github.com/honeycombio/libhoney-go"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	log = logrus.WithFields(logrus.Fields{
+		"source": "hny-lambda-ext-eventprocessor",
+	})
+	// ShutdownReasonFieldExtensionType is the field name for shutdown reason in shutdown reason event
+	ShutdownReasonFieldExtensionType = "lambda_extension.type"
+	// ShutdownReasonFieldRequestID is the field name used for request ID in shutdown reason event
+	ShutdownReasonFieldRequestID = "requestId"
+	// ShutdownReasonFieldInvokedFunctionARN is the field name used for function arn in shutdown reason event
+	ShutdownReasonFieldInvokedFunctionARN = "invokedFunctionArn"
+)
+
+// eventPoller is the interface that provides a next event for the event processor
+type eventPoller interface {
+	NextEvent(ctx context.Context) (*extension.NextEventResponse, error)
+}
+
+// eventFlusher is the interface that provides a way to create new libhoney events and flush them
+type eventFlusher interface {
+	NewEvent() *libhoney.Event
+	Flush()
+}
+
+// Server represents a server that polls and processes Lambda extension events
+type Server struct {
+	extensionClient    eventPoller
+	libhoneyClient     eventFlusher
+	invokedFunctionARN string
+	lastRequestId      string
+}
+
+// New takes an eventPoller and eventFlusher and returns a Server
+func New(extensionClient eventPoller, libhoneyClient eventFlusher) *Server {
+	return &Server{
+		extensionClient: extensionClient,
+		libhoneyClient:  libhoneyClient,
+	}
+}
+
+// Run executes an event loop to poll and process events from the Lambda extension API
+func (s *Server) Run(ctx context.Context, cancel context.CancelFunc) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			s.pollEventAndProcess(ctx, cancel)
+		}
+	}
+}
+
+// pollEventAndProcess polls the Lambda extension next event API and processes a single event
+func (s *Server) pollEventAndProcess(ctx context.Context, cancel context.CancelFunc) {
+	// Poll for event
+	res, err := s.extensionClient.NextEvent(ctx)
+	if err != nil {
+		log.WithError(err).Warn("Error from NextEvent")
+		return
+	}
+
+	// Ensure a flush happens and cancel is called if its a shutdown event
+	defer func() {
+		s.libhoneyClient.Flush()
+		if res.EventType == extension.Shutdown {
+			cancel()
+		}
+	}()
+
+	// Handles event types
+	switch eventType := res.EventType; eventType {
+	case extension.Invoke:
+		log.Debug("Received INVOKE event.")
+		s.lastRequestId = res.RequestID
+		s.invokedFunctionARN = res.InvokedFunctionARN
+	case extension.Shutdown:
+		log.Debug("Received SHUTDOWN event.")
+		if res.ShutdownReason != extension.ShutdownReasonSpindown && s.lastRequestId != "" {
+			log.WithField("res.ShutdownReason", res.ShutdownReason).Debug("Sending shutdown reason")
+			s.sendShutdownReason(res.ShutdownReason)
+		}
+	default:
+		log.WithField("res", res).Debug("Received unknown event")
+	}
```

Comment:
I don't think we need to do the flush and cancel with a deferred func. I think it would clearer if we called `Flush()` directly, then `cancel()` as part of the switch below in the Shutdown case. 

```suggestion
	// flush pending events to Honeycomb
	s.libhoneyClient.Flush()

	// Handles event types
	switch eventType := res.EventType; eventType {
	case extension.Invoke:
		log.Debug("Received INVOKE event.")
		s.lastRequestId = res.RequestID
		s.invokedFunctionARN = res.InvokedFunctionARN
	case extension.Shutdown:
		log.Debug("Received SHUTDOWN event.")
		if res.ShutdownReason != extension.ShutdownReasonSpindown && s.lastRequestId != "" {
			log.WithField("res.ShutdownReason", res.ShutdownReason).Debug("Sending shutdown reason")
			s.sendShutdownReason(res.ShutdownReason)
		}
		cancel()
	default:
		log.WithField("res", res).Debug("Received unknown event")
	}
```

Reply:
@MikeGoldsmith - thanks for taking a look at this! Yeah I agree, the deferred here is a bit wonky, what I was attempting to do with it though was avoid needing to flush twice. In your suggested commit the shutdown event isn't getting flushed. If I add Flush to the shutdown section in this example then we'd be flushing twice. But yeah I'm open to whatever makes sense here.